nixos/modules/system/boot/loader/grub/grub.nix

```diff
@@ -137,6 +138,67 @@ in
         '';
       };
 
+      users = mkOption {
+        default = {};
+        example = {
+          root = { password_pbkdf2_file = "/path/to/file"; };
+        };
+        description = ''
+          User accounts for GRUB. When specified, the GRUB command line and
+          all boot options except the default are password-protected.
+          All passwords and hashes provided will be stored in /boot/grub/grub.cfg,
+          and will be visible to any local user who can read this file. Additionally,
+          any passwords and hashes provided directly in a Nix configuration
+          (as opposed to external files) will be copied into the Nix store, and
+          will be visible to all local users.
+        '';
+        type = with types; attrsOf (submodule {
+          options = {
+            password_pbkdf2_file = mkOption {
+              example = "/path/to/file";
+              default = null;
+              type = with types; uniq (nullOr str);
+              description = ''
+                Specifies the path to a file containing the password hash
+                for the account, generated with grub-mkpasswd-pbkdf2.
+                This hash will be stored in /boot/grub/grub.cfg, and will
+                be visible to any local user who can read this file.
+              '';
+            };
+            password_pbkdf2 = mkOption {
+              example = "grub.pbkdf2.sha512.10000.674DFFDEF76E13EA...2CC972B102CF4355";
+              default = null;
+              type = with types; uniq (nullOr str);
+              description = ''
+                Specifies the password hash for the account,
+                generated with grub-mkpasswd-pbkdf2.
+                This hash will be copied to the Nix store, and will be visible to all local users.
+              '';
+            };
+            password_plaintext_file = mkOption {
```

Comment:
In retrospect, I wonder if the ability to specify credentials inline is unnecessary. It's extra code paths to test, and it's possible nobody would use it (I probably wouldn't).

Additionally, the perl script can just check if hashes start with "grub.pbkdf2". This would leave room for future hash formats to be supported under the same single option, and get rid of the ugly "pbkdf2" in the option name.

That would leave `passwordFile` and `hashedPasswordFile`, which sounds fine to me.

Reply:
> In retrospect, I wonder if the ability to specify credentials inline is unnecessary

Security-wise these options should be avoided but they are convenient sometimes: in NixOS tests for examples.

> Additionally, the perl script can just check if hashes start with "grub.pbkdf2"...

This looks like a good idea.

Reply:
Changes added. When I get a chance to spin up an VM I'll test everything.